### PR TITLE
feat: extHost log format refactor

### PR DESCRIPTION
**Pull request by faga295**
### Types

- [x] 🎉 New Features
- [ ] 🐛 Bug Fixes
- [ ] 📚 Documentation Changes
- [ ] 💄 Code Style Changes
- [ ] 💄 Style Changes
- [ ] 🪚 Refactors
- [ ] 🚀 Performance Improvements
- [ ] 🏗️ Build System
- [ ] ⏱ Tests
- [ ] 🧹 Chores
- [ ] Other Changes

### Background or solution
#2816
before:
image

after:
image

### <samp>🤖 Generated by Copilot at dfca83a</samp>

*  Add RPC functionality for logging messages from the server side to the client side via the `ILogServiceForClient` service ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-7dea0fd7105dc76ae2efb54ce4f8351bb2665805ae1d6eefc1aab3b3019f2333R280), [link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-7dea0fd7105dc76ae2efb54ce4f8351bb2665805ae1d6eefc1aab3b3019f2333R355-R360))
   * Define the `clientLog` method in the `IRPCLogService` interface in `packages/logs-core/src/node/log.service.ts` ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-7dea0fd7105dc76ae2efb54ce4f8351bb2665805ae1d6eefc1aab3b3019f2333R280))
   * Implement the `clientLog` method in the `LogServiceForClient` class in `packages/logs-core/src/node/log.service.ts` ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-7dea0fd7105dc76ae2efb54ce4f8351bb2665805ae1d6eefc1aab3b3019f2333R355-R360))
* Add service definition and implementation for logging messages from different namespaces to the client side via the `LogManage` class ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-a41cfad7bf17e19132249ddb239ca12cde91edc96b4d4c909d5b9ddcdc35ea7fR16), [link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-7b5f365639312b02977bd28ab6a7ae70ec603d26fc7687bd2570e332d189ef85R49-R52))
   * Declare the `clientLog` method in the `ILogServiceForClient` interface in `packages/logs-core/src/common/log-core.ts` ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-a41cfad7bf17e19132249ddb239ca12cde91edc96b4d4c909d5b9ddcdc35ea7fR16))
   * Define the `clientLog` method in the `LogManage` class in `packages/logs-core/src/browser/log-manage.ts` ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-7b5f365639312b02977bd28ab6a7ae70ec603d26fc7687bd2570e332d189ef85R49-R52))
* Add dependency injection and feature enhancement for logging messages from the extension host process to the client side via the `ExtensionNodeServiceImpl` class ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-96a7edbadf689ce6f8bd685ced0e2a3868a623e7ee060b6692a440d74490dac1R38), [link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-96a7edbadf689ce6f8bd685ced0e2a3868a623e7ee060b6692a440d74490dac1R66-R68), [link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-96a7edbadf689ce6f8bd685ced0e2a3868a623e7ee060b6692a440d74490dac1L376-R383))
   * Import the `ILogServiceForClient` interface from the `@opensumi/ide-logs` package to `packages/extension/src/node/extension.service.ts` ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-96a7edbadf689ce6f8bd685ced0e2a3868a623e7ee060b6692a440d74490dac1R38))
   * Add the `logService` property of type `ILogServiceForClient` to the `ExtensionNodeServiceImpl` class and annotate it with the `@Autowired` decorator ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-96a7edbadf689ce6f8bd685ced0e2a3868a623e7ee060b6692a440d74490dac1R66-R68))
   * Call the `clientLog` method of the `logService` property in the `handleOutput` method of the `ExtensionNodeServiceImpl` class ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-96a7edbadf689ce6f8bd685ced0e2a3868a623e7ee060b6692a440d74490dac1L376-R383))
* Replace the `arguments` object with the rest parameter syntax `...args` in the `_wrapConsoleMethod` function in `packages/extension/src/hosted/ext.process-base.ts` ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-c7d2dcfbc69eca7fdb8b6a13352db78e0cd7c3f8ab2d719412bb0541d6577a30L133-R134))
* Remove a trailing whitespace in the interface `ExtProcessConfig` in `packages/extension/src/hosted/ext.process-base.ts` ([link](https://github.com/opensumi/core/pull/2966/files?diff=unified&w=0#diff-c7d2dcfbc69eca7fdb8b6a13352db78e0cd7c3f8ab2d719412bb0541d6577a30L73-R77))

### <samp>🤖 Generated by Copilot at dfca83a</samp>

This pull request adds a feature to log extension host output messages to the client side using the `ILogServiceForClient` service. It defines and implements the `clientLog` method in the `ILogServiceForClient` interface and its subclasses, and uses it in the `ExtensionNodeServiceImpl` class. It also improves the code style of the `ExtProcessConfig` interface and the `_wrapConsoleMethod` function.

### CODE REVIEW

File: packages/logs-core/src/node/log.service.ts [hunk 1]
```diff
@@ -351,6 +352,12 @@ export class LogServiceForClient extends RPCService<IRPCLogService> implements I
     return this.loggerManager.getLogFolder();
   }
 
+  clientLog(namespace: SupportLogNamespace, message: string) {
+    if (this.client) {
+      this.client.clientLog(namespace, message);
+    }
+  }
+
```
Comment (faga295) [inline packages/logs-core/src/node/log.service.ts:360]: @erha19 I want the backend call frontend use this function like `todoListExample`, but `this.client` always be `undefined`, it confused me for a long time

Comment (erha19) [reply packages/logs-core/src/node/log.service.ts:360]: @faga295 目前 OpenSumi 内的服务模型还是 1 对 1 的 `Node <-> Browser` 通信结构，以 WebIDE 场景为例，如果你前端页面没有连接上，这里的 `this.client` 确实是会 `undefined` 的。

另外，开发阶段你也可以在前后端 Log 一些日志调试看一下执行流程，调试技巧看这里：[如何调试](https://opensumi.com/zh/docs/develop/how-to-contribute#%E8%B0%83%E8%AF%95)

Comment (faga295) [reply packages/logs-core/src/node/log.service.ts:360]: @erha19 https://github.com/opensumi/core/blob/eefb8745ff449165c7f1e92f14af26f6a74f801a/packages/core-node/src/connection.ts#L90-L140 不是很能够理解为什么需要这个childInjector，我看他返回的`serviceChildInjector`后面也没有用上，我尝试着改了一下，直接给injector addProviders
像这样
image
最终可以实现
image
没有这个改动，`this.client`一直是undefined

Comment (faga295) [reply packages/logs-core/src/node/log.service.ts:360]: 看样子在设计上backService和Provider的injector就是分离的，所以我在` @Autowired(ILogServiceForClient)`的时候并不会获取到处理过`instance.rpcClient`的那个实例，而是新建一个，不知道怎么样可以让`@Autowired(ILogServiceForClient)`获得带有`rpcClient`的实例

Comment (erha19) [reply packages/logs-core/src/node/log.service.ts:360]: @faga295 目前框架内需要遵循一定的约定去进行前后端的通信，即后端服务需要 `extends RPCService` ，这个暂时还是不支持通过 `@Autowired(ILogServiceForClient)` 去获取实例，这点对于你完成这个课题应该不会有太大影响... 搞清楚前后端通信如何写就行了，关于通信层后续还会有新的改造。

这个课题的核心任务还是需要重新梳理/完善一下目前框架内的 Log 逻辑，在此基础上，支持插件进程的日志输出。

**Comment by erha19** [comment]
@faga295 https://github.com/opensumi/core/pull/2966#issuecomment-1656117950 点击这里的链接签署一下开源的 CLA 协议

File: packages/extension/src/hosted/ext.process-base.ts [hunk 2]
```diff
@@ -127,11 +128,31 @@ function _wrapConsoleMethod(method: 'log' | 'info' | 'warn' | 'error') {
...
 
+  // eslint-disable-next-line no-console
   Object.defineProperty(console, method, {
     set: () => {},
     get: () =>
-      function () {
-        original(arguments);
+      function (...args: any[]) {
+        // eslint-disable-next-line no-console
+        // console.log('wrapconsole',...args);
```
Comment (erha19) [inline packages/extension/src/hosted/ext.process-base.ts:137]: 多余的代码直接删了就行了

File: packages/extension/src/hosted/ext.process-base.ts [hunk 3]
```diff
@@ -127,11 +128,31 @@ function _wrapConsoleMethod(method: 'log' | 'info' | 'warn' | 'error') {
...
+          original(...args);
+        } else {
+          const methodColorMap = {
+            info: 'green',
+            log: 'green',
+            warn: 'yellow',
+            error: 'red',
+          };
+          const color = methodColorMap[method];
+          original(getColor(color, `[extension:${method}]`), ...args);
```
Comment (erha19) [inline packages/extension/src/hosted/ext.process-base.ts:154]: 不建议在 Log 日志中追加这些 ASCII 颜色，后续做日志分析及前后端统一/格式化都会增加额外成本，可以参考一下 VS Code 内的这部分分层的实现，统一前/后(包含 NodeJS 及插件进程) 的基础实现，再根据不同场景去做自定义输出。

https://github.com/microsoft/vscode/blob/24c44070aefd4d14a9e7b8f1dbd4572b13c83bbd/src/vs/platform/log/common/log.ts#L268

**Review by erha19 — CHANGES_REQUESTED** [review]
另外，修改前后效果可以贴一个对比图看下，再改改

File: packages/core-common/src/log.ts [hunk 4]
```diff
@@ -202,6 +202,40 @@ export interface IDebugLog extends ICoreLogger {
  */
 const isNode = typeof process !== 'undefined' && process.release;
 const isChrome = !isNode && /Chrome/.test(navigator.userAgent) && /Google Inc/.test(navigator.vendor);
+
+export function getColor(color: string, message: string) {
+  if (!isNode && !isChrome) {
+    return message;
+  }
+  const colors = {
```
Comment (erha19) [inline packages/core-common/src/log.ts:210]: 这里还需要考虑最终数据落盘的情况，及写入到日志文件中的内容，如默认日志路径为 `~/.sumi/logs/`, 写入的逻辑见：

Node 层写入逻辑见：
https://github.com/opensumi/core/blob/58b998d9e1f721928f576579f16ded46b7505e84/packages/logs-core/src/node/logger.ts#L16

插件层写入逻辑见：
https://github.com/opensumi/core/blob/2c8e2a0281a2f36d35311eb4897c1c082a22f07b/packages/extension-storage/src/browser/storage.ts#L187

这里面存在比较多的写法不规范，可读性差的问题可以一并处理一下

**Review by erha19 — CHANGES_REQUESTED** [review]

File: packages/core-common/src/log.ts (hunk 4 above)
Comment (faga295) [reply packages/core-common/src/log.ts:210]: 数据落盘的话是插件进程和插件的输出都要落盘吗，因为我看之前的逻辑没有落盘

File: packages/extension/src/hosted/ext.process-base.ts (hunk 3 above)
Comment (faga295) [reply packages/extension/src/hosted/ext.process-base.ts:154]: 按照目前这样的方式写的话(在Object.defineProperty中), 用封装后的logger进行输出，会有无限循环的get console.*的问题，因为logger也会去get console.*

Comment (erha19) [reply packages/extension/src/hosted/ext.process-base.ts:154]: 这里顶层的 logger 应该就是 console.log，插件进程内是要避免一些循环引用，上面的建议是针对所有的 logger 结构

File: packages/core-common/src/log.ts (hunk 4 above)
Comment (erha19) [reply packages/core-common/src/log.ts:210]: 插件日志也是有落盘情况的，一般集成侧会通过自己定制 LogClass 的方式去捕获这块日志落盘

File: packages/extension/src/hosted/ext.process-base.ts (hunk 3 above)
Comment (erha19) [reply packages/extension/src/hosted/ext.process-base.ts:154]: 这个 PR 可以先不做这层改造，精简一下修改内容，分 PR 实现，别在原始输出日志中增加输出格式（颜色，前缀等）

File: packages/logs-core/src/node/log.service.ts [hunk 5]
```diff
@@ -351,6 +352,12 @@ export class LogServiceForClient extends RPCService<IRPCLogService> implements I
     return this.loggerManager.getLogFolder();
   }
 
+  clientLog(namespace: SupportLogNamespace, message: string) {
```
Comment (erha19) [inline packages/logs-core/src/node/log.service.ts:355]: 前端输出日志应该仅限于插件日志，同时只在开发环境下输出，这里这样写会影响生产环境下的使用，同时也缺乏控制力度，问题二的解决建议另外开个 PR，这个 PR 主要把内容格式梳理一下

**Review by erha19 — CHANGES_REQUESTED** [review]

File: packages/core-common/src/log.ts (hunk 4 above)
Comment (faga295) [reply packages/core-common/src/log.ts]: > 这里面存在比较多的写法不规范，可读性差的问题可以一并处理一下

这里的写法不规范指的是什么，从写入的日志文件没有看出来

Comment (erha19) [reply packages/core-common/src/log.ts]: 主要是代码部分，如果有些地方发现理解的与命名/逻辑有冲突，直接改就行了 @faga295

File: packages/extension/src/hosted/api/vscode/ext.host.command.ts [hunk 6]
```diff
@@ -67,7 +68,9 @@ export function createCommandsApiFactory(
       return extHostCommands.registerCommand(true, id, (...args: any[]): any => {
         const activeTextEditor = extHostEditors.activeEditor ? extHostEditors.activeEditor.textEditor : undefined;
         if (!activeTextEditor) {
-          getDebugLogger().warn('Cannot execute ' + id + ' because there is no active text editor.');
+          getDebugLogger(SupportLogNamespace.ExtensionHost).warn(
```
Comment (erha19) [inline packages/extension/src/hosted/api/vscode/ext.host.command.ts:71]: 这里更合适的写法实在顶部的 `ext.host.api.impl` 中做好 logger 的初始化，然后再传入各个实现中，减少多次初始化的问题

File: packages/extension/src/node/extension.service.ts [hunk 7]
```diff
@@ -371,11 +370,7 @@ export class ExtensionNodeServiceImpl implements IExtensionNodeService {
         this.onDidSetInspectPort.fire();
       } else {
         // 输出插件进程日志
-        if (output.type === OutputType.STDERR) {
-          this.extHostLogger.error(util.format(output.data, ...output.format));
-        } else {
-          this.extHostLogger.log(util.format(output.data, ...output.format));
```
Comment (erha19) [inline packages/extension/src/node/extension.service.ts:377]: 下面这个 else 判断为啥删除了？

**Review by erha19 — CHANGES_REQUESTED** [review]

**Comment by faga295** [comment]
@erha19 cc

File: packages/extension/src/node/extension.service.ts [hunk 8]
```diff
@@ -62,7 +62,7 @@ export class ExtensionNodeServiceImpl implements IExtensionNodeService {
   @Autowired(INodeLogger)
   private readonly logger: INodeLogger;
 
-  private readonly extHostLogger = getDebugLogger(SupportLogNamespace.ExtensionHost);
+  private readonly extHostLogger = console;
```
Comment (erha19) [inline packages/extension/src/node/extension.service.ts:65]: 这里实际上是从 Node 层实际输出插件进程日志的位置，也是最适合做输出格式化的地方，你在 ext.host.api.impl.ts 做格式化的时机有点早了，那里面建议直接使用 console，或者是 getDebugLogger, 如 :

```diff
- const debugLogger = getDebugLogger(SupportLogNamespace.ExtensionHost);
+ const logger = console;
```
然后在这里再做输出的格式化

**Review by erha19 — CHANGES_REQUESTED** [review]

File: packages/extension/src/node/extension.service.ts (hunk 8 above)
Comment (faga295) [reply packages/extension/src/node/extension.service.ts:65]: 噢噢 明白了 extHost正常console输出，在extension.service 去加前缀[extHost:xxx]

Comment (faga295) [reply packages/extension/src/node/extension.service.ts:65]: 那也不用在ext.host.api.impl里去初始化logger了， 直接把直接getDebugLogger用console代替了

Comment (faga295) [reply packages/extension/src/node/extension.service.ts:65]: 但是这样的话 没办法区分 warn和log

Comment (erha19) [reply packages/extension/src/node/extension.service.ts:65]: 确实有这个问题，那日志输出的时候可以在插件进程做一层格式化如 :

`[extHost:log]` -> `[Log]`

仅保留日志的类型，然后再在外部 Node 层输出时，包裹上 

`[Extension Host] ...` 前面这种结构，最终输出的日志信息就会比较清楚了，如：

```
[Extension Host] [LOG] ....
```

**Review by erha19 — CHANGES_REQUESTED** [review]

**Comment by faga295** [comment]
这样也行，不过这样会和其他的日志格式不太一致, 其他的日志输出是[${namespace}:${level}]的形式， extHost本身也是一个namespace， 最终的输出我认为和之前的日志输出保持一致会更合适一些

**Comment by faga295** [comment]
@erha19 cc

**Pull request closed without merging**